
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Taxonomic Classification Walkthrough &#8212; dahak-taco  documentation</title>
    <link rel="stylesheet" href="../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <script type="text/javascript" src="../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Assembly Walkthrough" href="assembly.html" />
    <link rel="prev" title="Snakemake Rules" href="../intro/snakemakerules.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  <div class="document">
    
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Taxonomic Classification Walkthrough</a><ul>
<li><a class="reference internal" href="#listing-rules">Listing Rules</a></li>
<li><a class="reference internal" href="#rule-pull-biocontainers">Rule: Pull Biocontainers</a></li>
<li><a class="reference internal" href="#rule-download-and-unpack-sourmash-sbts">Rule: Download and Unpack Sourmash SBTs</a></li>
<li><a class="reference internal" href="#rule-calculate-signatures-default-parameters">Rule: Calculate Signatures (Default Parameters)</a></li>
<li><a class="reference internal" href="#rule-calculate-signatures-modified-parameters">Rule: Calculate Signatures (Modified Parameters)</a></li>
<li><a class="reference internal" href="#list-of-rules">List of Rules</a></li>
<li><a class="reference internal" href="#list-of-rule-files">List of Rule Files</a></li>
</ul>
</li>
</ul>
<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
      <li>Previous: <a href="../intro/snakemakerules.html" title="previous chapter">Snakemake Rules</a></li>
      <li>Next: <a href="assembly.html" title="next chapter">Assembly Walkthrough</a></li>
  </ul></li>
</ul>
</div>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/workflows/taxclass.md.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="taxonomic-classification-walkthrough">
<span id="taxonomic-classification-walkthrough"></span><h1>Taxonomic Classification Walkthrough<a class="headerlink" href="#taxonomic-classification-walkthrough" title="Permalink to this headline">¶</a></h1>
<p>In this document, we walk through a few
rules in the taxonomic workflow
to illustrate how to run dahak workflows
using custom rules and custom workflow
parameters.</p>
<p>See <a class="reference external" href="https://github.com/dahak-metagenomics/dahak/tree/master/workflows/taxonomic_classification">dahak - taxonomic classification workflow</a>.</p>
<div class="section" id="listing-rules">
<span id="listing-rules"></span><h2>Listing Rules<a class="headerlink" href="#listing-rules" title="Permalink to this headline">¶</a></h2>
<p>Start by selecting which rule from the workflow you want to run.
To see available rules, run:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ ./taco ls
</pre></div>
</div>
<p>Add the rule you want to run to the workflow configuration file.</p>
<p>Check the relevant <code class="docutils literal notranslate"><span class="pre">rules/dahak/*.settings</span></code> file for any settings
to change in the workflow parameters file.</p>
</div>
<div class="section" id="rule-pull-biocontainers">
<span id="rule-pull-biocontainers"></span><h2>Rule: Pull Biocontainers<a class="headerlink" href="#rule-pull-biocontainers" title="Permalink to this headline">¶</a></h2>
<p>Start with a simple example of a single workflow step
with no input or output files. To run the
<code class="docutils literal notranslate"><span class="pre">pull_biocontainers</span></code> workflow:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cat pull-biocontainers-workflow.json
{
    &quot;workflow_target&quot; : &quot;pull_biocontainers&quot;
}
</pre></div>
</div>
<p>Now run the workflow (include <code class="docutils literal notranslate"><span class="pre">-n</span></code> to do a dry run):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ ./taco -n pull-biocontainers-workflow # dry run

$ ./taco pull-biocontainers-workflow
</pre></div>
</div>
<p>Note that the <code class="docutils literal notranslate"><span class="pre">biocontainers</span></code> rule is defined in
<code class="docutils literal notranslate"><span class="pre">rules/dahak/biocontainers.rule</span></code>
and the corresponding default parameter
values are defined in <code class="docutils literal notranslate"><span class="pre">rules/dahak/biocontainers.settings</span></code>.</p>
<p>If we wanted to modify any parameters,
we could add configuration parameters
to <code class="docutils literal notranslate"><span class="pre">pull-biocontainers-params.json</span></code>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cat pull-biocontainers-params.json
{
    &#39;biocontainers&#39; : {
        &#39;sourmash&#39; : {
            &#39;version&#39; : &#39;2.0.0a2--py36_0&#39;
        }
    }
}
</pre></div>
</div>
<p>Now run the workflow, specifying both the workflow
and parameter configuration files on the command line
(include <code class="docutils literal notranslate"><span class="pre">-n</span></code> to do a dry run):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ ./taco -n pull-biocontainers-workflow pull-biocontainers-params # dry run

$ ./taco pull-biocontainers-workflow pull-biocontainers-params
</pre></div>
</div>
</div>
<div class="section" id="rule-download-and-unpack-sourmash-sbts">
<span id="rule-download-and-unpack-sourmash-sbts"></span><h2>Rule: Download and Unpack Sourmash SBTs<a class="headerlink" href="#rule-download-and-unpack-sourmash-sbts" title="Permalink to this headline">¶</a></h2>
<p>Like the prior step, this step has no
input or output file names to set.</p>
<p>This step downloads SBTs containing hashes
from the microbial genomes in the NCBI Databank
and RefSeq databases.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cat unpack-sbt-workflow.json
{
    &quot;workflow_target&quot; : &quot;unpack_sourmash_sbts&quot;
}

$ ./taco -n unpack-sbt-workflow # dry run

$ ./taco unpack-sbt-workflow
</pre></div>
</div>
<p>This calls the <code class="docutils literal notranslate"><span class="pre">unpack_sourmash_sbts</span></code> rule,
which calls the <code class="docutils literal notranslate"><span class="pre">download_sourmash_sbts</span></code> rule.
These rules are defined in <code class="docutils literal notranslate"><span class="pre">rules/dahak/sourmash_sbt.rule</span></code>.</p>
</div>
<div class="section" id="rule-calculate-signatures-default-parameters">
<span id="rule-calculate-signatures-default-parameters"></span><h2>Rule: Calculate Signatures (Default Parameters)<a class="headerlink" href="#rule-calculate-signatures-default-parameters" title="Permalink to this headline">¶</a></h2>
<p>The calculate signatures rule can be run with default values:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cat calc-sigs.json
{
    &quot;workflow_target&quot; : &quot;calculate_signatures&quot;
}

$ ./taco -n calc-sigs # dry run

$ ./taco calc-sigs
</pre></div>
</div>
</div>
<div class="section" id="rule-calculate-signatures-modified-parameters">
<span id="rule-calculate-signatures-modified-parameters"></span><h2>Rule: Calculate Signatures (Modified Parameters)<a class="headerlink" href="#rule-calculate-signatures-modified-parameters" title="Permalink to this headline">¶</a></h2>
<p>We can modify parameters for the calculate signatures workflow.</p>
<p>If we examine the <code class="docutils literal notranslate"><span class="pre">calculate_signatures</span></code> rule file at
<code class="docutils literal notranslate"><span class="pre">rules/dahak/calculate_signatures.rule</span></code> we see that three
settings files are used:</p>
<ul class="simple">
<li>read settings (group parameters for any task involving reading sequences)</li>
<li>calculate signature settings (rule-specific settings for calculating signatures)</li>
<li>biocontainer settings (container URLs and versions)</li>
</ul>
<p>We can override any parameters in those files
to change how the <code class="docutils literal notranslate"><span class="pre">calculate_signatures</span></code> workflow
works.</p>
<p><strong>Example 1:</strong> Change the name of the forward and reverse read files.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cat calc-sigs-params.json
{
    &quot;reads&quot; : {
        &quot;fq_fwd&quot; : &quot;{base}_1.trim{ntrim}.fq.gz&quot;,
        &quot;fq_rev&quot; : &quot;{base}_2.trim{ntrim}.fq.gz&quot;
    }
}

$ cat calc-sigs.json
{
    &quot;workflow_target&quot; : &quot;calculate_signatures&quot;
}

$ ./taco -n calc-sigs calc-sigs-params # dry run

$ ./taco calc-sigs calc-sigs-params
</pre></div>
</div>
<p><strong>Example 2:</strong> Change the k values used to calculate signatures.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cat calc-sigs-params.json
{
    &quot;calculate_signatures&quot; : {
        &quot;kvalues&quot; : [21, 31, 51, 101]
    }
}

$ cat calc-sigs.json
{
    &quot;workflow_target&quot; : &quot;calculate_signatures&quot;
}

$ ./taco -n calc-sigs calc-sigs-params # dry run

$ ./taco calc-sigs calc-sigs-params
</pre></div>
</div>
<p><strong>Example 3:</strong> Change the sequence and merge file naming schema.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cat calc-sigs-params.json
{
    &quot;calculate_signatures&quot; : {
        &quot;sig_name&quot; : &quot;{base}.trim{ntrim}.scaled{scale}.k{kvalues_fname}.sig&quot;
    }
}
</pre></div>
</div>
<p>NOTE: <code class="docutils literal notranslate"><span class="pre">kvalues_fname</span></code> is a variable created when the rule is run,
and is the result of <code class="docutils literal notranslate"><span class="pre">&quot;_&quot;.join(config['calculate_signatures']['kvalues'])</span></code>.</p>
<p>TODO: improve the abstraction.</p>
<p><strong>Example 4:</strong> Implement all three of the above parameter changes.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cat calc-sigs-params.json
{
    &quot;reads&quot; : {
        &quot;fq_fwd&quot; : &quot;{base}_1.trim{ntrim}.fq.gz&quot;,
        &quot;fq_rev&quot; : &quot;{base}_2.trim{ntrim}.fq.gz&quot;
    },
    &quot;calculate_signatures&quot; : {
        &quot;kvalues&quot; : [21, 31, 51, 101],
        &quot;sig_name&quot; : &quot;{base}.trim{ntrim}.scaled{scale}.k{kvalues_fname}.sig&quot;
    }
}

$ cat calc-sigs.json
{
    &quot;workflow_target&quot; : &quot;calculate_signatures&quot;
}

$ ./taco -n calc-sigs calc-sigs-params # dry run

$ ./taco calc-sigs calc-sigs-params
</pre></div>
</div>
</div>
<div class="section" id="list-of-rules">
<span id="list-of-rules"></span><h2>List of Rules<a class="headerlink" href="#list-of-rules" title="Permalink to this headline">¶</a></h2>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>pull_biocontainers
    
    - Pull the latest version of sourmash, kaiju, and krona
    - Version numbers are set in biocontainers.settings
    - To call this rule, ask for the file .pulled_containers
    
download_sourmash_sbts
    
    Downoad the sourmash SBTs from spacegraphcats

    To call this rule, request sourmash SBT json file for the specified database.
    
unpack_sourmash_sbts
    
    Unpack the sourmash SBTs
    
calculate_signatures
    
    Calculate signatures from trimmed data using sourmash
    
download_trimmed_data
    
    Download the trimmed data from OSF

    To call this rule, request the files listed in trimmed_data.dat
    
unpack_kaiju
    
    Download and unpack the kaiju database.
    
run_kaiju
    
    Run kaiju
    
kaiju2krona
    
    Convert kaiju results to krona results,
    and generate a report.
    
kaiju2kronasummary
    
    Convert kaiju results to krona results,
    and generate a report.
    
filter_taxa_total
    
    Filter out taxa with low abundances by obtaining genera that
    comprise at least {pct} percent of the total reads
    (default: 1%)
    
filter_taxa_class
    
    For comparison, take the genera that comprise
    at least {pct} percent of all of the classified reads
    (default: 1%)
    
visualize_krona
    
    Visualize the results of the
    full and filtered taxonomic
    classifications using krona.
</pre></div>
</div>
</div>
<div class="section" id="list-of-rule-files">
<span id="list-of-rule-files"></span><h2>List of Rule Files<a class="headerlink" href="#list-of-rule-files" title="Permalink to this headline">¶</a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">rules</span><span class="o">/</span>
    <span class="n">dahak</span><span class="o">/</span>
        <span class="n">biocontainers</span><span class="o">.</span><span class="n">rule</span>
        <span class="n">calculate_signatures</span><span class="o">.</span><span class="n">rule</span>
        <span class="n">filter_taxa</span><span class="o">.</span><span class="n">rule</span>
        <span class="n">kaiju</span><span class="o">.</span><span class="n">rule</span>
        <span class="n">kaiju2krona</span><span class="o">.</span><span class="n">rule</span>
        <span class="n">krona_visualization</span><span class="o">.</span><span class="n">rule</span>
        <span class="n">sourmash_sbt</span><span class="o">.</span><span class="n">rule</span>
        <span class="n">trimmed_data</span><span class="o">.</span><span class="n">rule</span>
</pre></div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
    <div class="clearer"></div>
  </div>
    <div class="footer">
      &copy;2018.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.7.2</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.10</a>
      
      |
      <a href="../_sources/workflows/taxclass.md.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/charlesreid1/dahak-taco" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>