
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Read Filtering Walkthrough &#8212; dahak-taco  documentation</title>
    <link rel="stylesheet" href="../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/bootstrap.min.css" type="text/css" />
    <script type="text/javascript" src="../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Workflow: read_filtering" href="../read_filtering/read_filtering.html" />
    <link rel="prev" title="Worker Node Setup Walkthrough" href="awssetup.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  <div class="document">
    
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Read Filtering Walkthrough</a><ul>
<li><a class="reference internal" href="#setup">Setup</a></li>
<li><a class="reference internal" href="#clone-the-repo">Clone The Repo</a></li>
<li><a class="reference internal" href="#start-the-workflow">Start the Workflow</a><ul>
<li><a class="reference internal" href="#step-1-rule-download-the-data">Step 1: Rule: Download the Data</a></li>
<li><a class="reference internal" href="#step-2-rule-pre-trim-quality-assessment">Step 2: Rule: Pre-Trim Quality Assessment</a></li>
<li><a class="reference internal" href="#step-3-rule-trim-reads">Step 3: Rule: Trim Reads</a></li>
<li><a class="reference internal" href="#step-4-rule-post-trim-quality-assessment">Step 4: Rule: Post-Trim Quality Assessment</a></li>
</ul>
</li>
</ul>
</li>
</ul>
<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
      <li>Previous: <a href="awssetup.html" title="previous chapter">Worker Node Setup Walkthrough</a></li>
      <li>Next: <a href="../read_filtering/read_filtering.html" title="next chapter">Workflow: <code class="docutils literal notranslate"><span class="pre">read_filtering</span></code></a></li>
  </ul></li>
</ul>
</div>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/walkthrus/readfilt.md.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="read-filtering-walkthrough">
<span id="read-filtering-walkthrough"></span><h1>Read Filtering Walkthrough<a class="headerlink" href="#read-filtering-walkthrough" title="Permalink to this headline">¶</a></h1>
<p>This walkthrough covers a read filtering workflow.</p>
<p>In this document we will:</p>
<ul class="simple">
<li>Download raw sequence data from OSF data store (wget).</li>
<li>Assess the quality of the pre-filtered reads (fastqc)</li>
<li>Trim and filter the reads based on a quality threshold (trimmomatic)</li>
<li>Assess the quality of the post-filtered reads (fastqc)</li>
</ul>
<p>This workflow specifically covers the use of taco.
See the <a class="reference external" href="https://github.com/dahak-metagenomics/dahak/tree/master/workflows/read_filtering">walkthroughs dir of the dahak repo</a>
for the original shell-based walkthrough.</p>
<div class="section" id="setup">
<span id="setup"></span><h2>Setup<a class="headerlink" href="#setup" title="Permalink to this headline">¶</a></h2>
<p>This node assumes you have the following software:</p>
<ul class="simple">
<li>Python</li>
<li>Conda</li>
<li>Snakemake</li>
<li>Singularity</li>
</ul>
<p>See <a class="reference external" href="/INSTALLING.md">INSTALLING</a> for more info.</p>
<p>If you do not have an environment set up, the
<a class="reference external" href="setup.md">Worker Node Setup Walkthrough</a>
covers setting up a worker node on AWS.</p>
</div>
<div class="section" id="clone-the-repo">
<span id="clone-the-repo"></span><h2>Clone The Repo<a class="headerlink" href="#clone-the-repo" title="Permalink to this headline">¶</a></h2>
<p>Start by cloning a copy of dahak-taco:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ git clone https://github.com/dahak-metagenomics/dahak-taco.git
$ cd dahak-taco/
</pre></div>
</div>
<p>As mentioned in <a class="reference external" href="/INSTALLING.md">INSTALLING</a>,
there is nothing to install for taco itself.</p>
</div>
<div class="section" id="start-the-workflow">
<span id="start-the-workflow"></span><h2>Start the Workflow<a class="headerlink" href="#start-the-workflow" title="Permalink to this headline">¶</a></h2>
<p>For the read filtering walkthrough we will be downloading
two <code class="docutils literal notranslate"><span class="pre">*.fq.gz</span></code> files, assessing their quality,
and trimming them based on the quality of the reads.</p>
<p>The workflow has four steps:</p>
<ul class="simple">
<li>Download the data</li>
<li>Assess the quality before trimming</li>
<li>Trim the data</li>
<li>Assess the quality after trimming</li>
</ul>
<div class="section" id="step-1-rule-download-the-data">
<span id="step-1-rule-download-the-data"></span><h3>Step 1: Rule: Download the Data<a class="headerlink" href="#step-1-rule-download-the-data" title="Permalink to this headline">¶</a></h3>
<p>Our three required inputs are:</p>
<ul class="simple">
<li>Workflow name (<code class="docutils literal notranslate"><span class="pre">read_filtering</span></code>)</li>
<li>Workflow configuration file (in <code class="docutils literal notranslate"><span class="pre">goodies/readfilt1config.json</span></code>)</li>
<li>Workflow parameters file (in <code class="docutils literal notranslate"><span class="pre">goodies/readfilt1params.json</span></code>)</li>
</ul>
<p>Here is the workflow configuration file we will use.
The workflow targets it specifies are two files with
unprocessed sequence reads.</p>
<p>(These files are the inputs to the rest of the
read filtering workflow.)</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cat goodies/readfilt1config.json
{
    &quot;short_description&quot;: &quot;Read Filtering Walkthrough 1 - Configuration&quot;,
    &quot;workflow_targets&quot; : [&quot;data/SRR606249_1.fq.gz&quot;,
                          &quot;data/SRR606249_2.fq.gz&quot;]
}
</pre></div>
</div>
<p>Next, we have our workflow parameters file.
This parameters file should define any parameters
used for the single step we’re executing.
In our case, we have to tell Snakemake
where to download the read files.
This parameters file defines a long list of files,
but rules are defined on a file-by-file basis
so we only downlod the files we need:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cat goodies/readfilt1params.json
{
    &quot;short_description&quot;: &quot;Read Filtering Walkthrough 1 Parameters&quot;,
    &quot;read_filtering&quot; : {
        &quot;read_patterns&quot; : {
            &quot;pre_trimming_pattern&quot;  : &quot;{sample}_{direction}.fq.gz&quot;,
        },
        &quot;read_files&quot; : {
            &quot;SRR606249_1.fq.gz&quot; :           &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f0f9156c613b026430dbc7&quot;,
            &quot;SRR606249_2.fq.gz&quot; :           &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f0fc7fb83f69026076be47&quot;,
            &quot;SRR606249_subset10_1.fq.gz&quot; :  &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f10134b83f69026377611b&quot;,
            &quot;SRR606249_subset10_2.fq.gz&quot; :  &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f101f26c613b026330e53a&quot;,
            &quot;SRR606249_subset25_1.fq.gz&quot; :  &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f1039a594d900263120c38&quot;,
            &quot;SRR606249_subset25_2.fq.gz&quot; :  &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f104ed594d90026411f486&quot;,
            &quot;SRR606249_subset50_1.fq.gz&quot; :  &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f1082d6c613b026430e5cf&quot;,
            &quot;SRR606249_subset50_2.fq.gz&quot; :  &quot;files.osf.io/v1/resources/dm938/providers/osfstorage/59f10ac6594d900262123e77&quot;
        }
    }
}
</pre></div>
</div>
<p>Both of these JSON files are contained in the
<code class="docutils literal notranslate"><span class="pre">goodies/</span></code> directory of the repo.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">read_patterns</span></code> parameter is optional and the
default value is shown here. This pattern is used
to create the download wildcard rule, so the files
that make up the <code class="docutils literal notranslate"><span class="pre">read_files</span></code> keys must match that
pattern or they will not be downloadable.</p>
<p>To run the workflow defined by this workflow,
config file, and parameters file, run taco as follows:</p>
<p>Dry run first with the <code class="docutils literal notranslate"><span class="pre">-n</span></code> flag:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ ./taco -n read_filtering \
    goodies/readfilt1config.json \
    goodies/readfilt1params.json
</pre></div>
</div>
<p>Then the real deal:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ ./taco -n read_filtering \
    goodies/readfilt1config.json \
    goodies/readfilt1params.json
</pre></div>
</div>
</div>
<div class="section" id="step-2-rule-pre-trim-quality-assessment">
<span id="step-2-rule-pre-trim-quality-assessment"></span><h3>Step 2: Rule: Pre-Trim Quality Assessment<a class="headerlink" href="#step-2-rule-pre-trim-quality-assessment" title="Permalink to this headline">¶</a></h3>
<p>The second step uses <code class="docutils literal notranslate"><span class="pre">fastq</span></code> to assess the
quality of the sequencer reads before any
trimming occurs.</p>
<p>The file suffix for files whose quality has been
assessed by fastqc is specified in the config file,
or can be left as the default <code class="docutils literal notranslate"><span class="pre">_fastqc</span></code>.</p>
<p>Here is the config file for the second step,
which is the pre-trim quality assessment.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cat goodies/readfilt2config.json
{
    &quot;short_description&quot;: &quot;Read Filtering Walkthrough 2 - Pre-Trim Quality Assessment - Configuration&quot;,
    &quot;workflow_targets&quot; : [&quot;data/SRR606249_1_fastqc.zip&quot;,
                          &quot;data/SRR606249_2_fastqc.zip&quot;]
}
</pre></div>
</div>
<p>While the second step of the workflow should use a
second config file, the parameters file should
include all the parameters from the prior step,
since Snakemake will still be running those workflow steps.</p>
<p>Here is the parameters file:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cat goodies/readfilt2params.json
{
    &quot;short_description&quot;: &quot;Read Filtering Walkthrough 2 - Pre-Trim Quality Assessment - Parameters&quot;,
    &quot;read_filtering&quot; : {
        &quot;quality_assessment&quot; : {
            &quot;fastqc_suffix&quot; : &quot;_fastqc&quot;
        },
        &quot;read_patterns&quot; : {
            &quot;pre_trimming_pattern&quot;  : &quot;{sample}_{direction}.fq.gz&quot;,
        },
        &quot;read_files&quot; : {
            ...
        }
    }
}
</pre></div>
</div>
<p>To run the workflow defined by this workflow,
config file, and parameters file, run taco as follows:</p>
<p>Dry run first with the <code class="docutils literal notranslate"><span class="pre">-n</span></code> flag:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ ./taco -n read_filtering \
    goodies/readfilt2config.json \
    goodies/readfilt2params.json
</pre></div>
</div>
<p>Then the real deal:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ ./taco read_filtering \
    goodies/readfilt2config.json \
    goodies/readfilt2params.json
</pre></div>
</div>
</div>
<div class="section" id="step-3-rule-trim-reads">
<span id="step-3-rule-trim-reads"></span><h3>Step 3: Rule: Trim Reads<a class="headerlink" href="#step-3-rule-trim-reads" title="Permalink to this headline">¶</a></h3>
<p>Step 3 uses trimmomatic to trim the reads based on quality.</p>
<p>This step requires us to output reads to a new trimmed filename.
We have to change the pre-trimming filename pattern,
which is too general: <code class="docutils literal notranslate"><span class="pre">{sample}_{direction}.fq.gz</span></code>
will match nearly every <code class="docutils literal notranslate"><span class="pre">post_trimming_pattern</span></code> we choose.</p>
<p>We should change the pattern parameters to include
either a prefix or a suffix:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>    <span class="s2">&quot;pre_trimming_pattern&quot;</span>  <span class="p">:</span>  <span class="s2">&quot;</span><span class="si">{sample}</span><span class="s2">_</span><span class="si">{direction}</span><span class="s2">_reads.fq.gz&quot;</span>
    <span class="s2">&quot;post_trimming_pattern&quot;</span>  <span class="p">:</span> <span class="s2">&quot;</span><span class="si">{sample}</span><span class="s2">_</span><span class="si">{direction}</span><span class="s2">_trim</span><span class="si">{qual}</span><span class="s2">.fq.gz&quot;</span>
</pre></div>
</div>
<p>Because we changed the target filenames for steps 1 and 2,
this will cause steps 1 and 2 to re-run.</p>
<p>This step also requires us to add an adapter file for
trimmomatic to use. This works like the read files:
we specify a target filename and a URL for the data.</p>
<p>The following step 3 config file explicitly specifies
targets for steps 1 and 2, in addition to step 3:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cat goodies/readfilt3config.json
{
    &quot;short_description&quot;: &quot;Read Filtering Walkthrough 2 - Pre-Trim Quality Assessment - Configuration&quot;,
    &quot;workflow_targets&quot; : [&quot;data/SRR606249_1_reads.fq.gz&quot;,
                          &quot;data/SRR606249_2_reads.fq.gz&quot;,
                          &quot;data/SRR606249_1_reads_fastqc.zip&quot;,
                          &quot;data/SRR606249_2_reads_fastqc.zip&quot;,
                          &quot;data/TruSeq2-PE.fa&quot;,
                          &quot;data/SRR606249_1_trim2.fq.gz&quot;,
                          &quot;data/SRR606249_2_trim2.fq.gz&quot;]
}
</pre></div>
</div>
<p>The parameters file contains updated parameters
(most options are not required, but this illustrates
what controls the user has over file names):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="p">{</span>
    <span class="s2">&quot;short_description&quot;</span><span class="p">:</span> <span class="s2">&quot;Read Filtering Walkthrough 3 - Trimming - Parameters&quot;</span><span class="p">,</span>
    <span class="s2">&quot;read_filtering&quot;</span> <span class="p">:</span> <span class="p">{</span>
        <span class="s2">&quot;quality_assessment&quot;</span> <span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;fastqc_suffix&quot;</span> <span class="p">:</span> <span class="s2">&quot;_fastqc&quot;</span>
        <span class="p">},</span>
        <span class="s2">&quot;quality_trimming&quot;</span> <span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;trim_suffix&quot;</span> <span class="p">:</span> <span class="s2">&quot;_se&quot;</span>
        <span class="p">},</span>
        <span class="s2">&quot;read_patterns&quot;</span> <span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;pre_trimming_pattern&quot;</span>  <span class="p">:</span> <span class="s2">&quot;</span><span class="si">{sample}</span><span class="s2">_</span><span class="si">{direction}</span><span class="s2">_reads.fq.gz&quot;</span><span class="p">,</span>
            <span class="s2">&quot;post_trimming_pattern&quot;</span> <span class="p">:</span> <span class="s2">&quot;</span><span class="si">{sample}</span><span class="s2">_</span><span class="si">{direction}</span><span class="s2">_trim</span><span class="si">{qual}</span><span class="s2">.fq.gz&quot;</span><span class="p">,</span>
            <span class="s2">&quot;adapter_pattern&quot;</span> <span class="p">:</span>       <span class="s2">&quot;</span><span class="si">{adapter}</span><span class="s2">.fa&quot;</span>
        <span class="p">},</span>
        <span class="s2">&quot;read_files&quot;</span> <span class="p">:</span> <span class="p">{</span>
            <span class="o">...</span>
        <span class="p">},</span>
        <span class="s2">&quot;adapter_files&quot;</span> <span class="p">:</span> <span class="p">{</span>
            <span class="s2">&quot;TruSeq2-PE.fa&quot;</span> <span class="p">:</span> <span class="s2">&quot;http://dib-training.ucdavis.edu.s3.amazonaws.com/mRNAseq-semi-2015-03-04/TruSeq2-PE.fa&quot;</span>
        <span class="p">}</span>
</pre></div>
</div>
<p>Dry run first with the <code class="docutils literal notranslate"><span class="pre">-n</span></code> flag:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ ./taco -n read_filtering \
    goodies/readfilt3config.json \
    goodies/readfilt3params.json
</pre></div>
</div>
<p>Then the real deal:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ ./taco read_filtering \
    goodies/readfilt3config.json \
    goodies/readfilt3params.json
</pre></div>
</div>
</div>
<div class="section" id="step-4-rule-post-trim-quality-assessment">
<span id="step-4-rule-post-trim-quality-assessment"></span><h3>Step 4: Rule: Post-Trim Quality Assessment<a class="headerlink" href="#step-4-rule-post-trim-quality-assessment" title="Permalink to this headline">¶</a></h3>
</div>
</div>
</div>


          </div>
        </div>
      </div>
    <div class="clearer"></div>
  </div>
    <div class="footer">
      &copy;2018.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.7.2</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.10</a>
      
      |
      <a href="../_sources/walkthrus/readfilt.md.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/charlesreid1/dahak-taco" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>